ddl/index.go

```diff
@@ -33,32 +34,71 @@ import (
 	"github.com/pingcap/tidb/util/types"
 )
 
-const maxPrefixLength = 767
+const maxPrefixLength = 3072
 
-func buildIndexInfo(tblInfo *model.TableInfo, unique bool, indexName model.CIStr,
-	idxColNames []*ast.IndexColName) (*model.IndexInfo, error) {
-	// build offsets
+func buildIndexColumns(columns []*model.ColumnInfo, idxColNames []*ast.IndexColName) ([]*model.IndexColumn, error) {
+	// build offsets.
 	idxColumns := make([]*model.IndexColumn, 0, len(idxColNames))
+
+	// the sum of length of all index columns.
+	sumLength := 0
+
 	for _, ic := range idxColNames {
-		col := findCol(tblInfo.Columns, ic.Column.Name.O)
+		col := findCol(columns, ic.Column.Name.O)
 		if col == nil {
 			return nil, errKeyColumnDoesNotExits.Gen("column does not exist: %s",
 				ic.Column.Name)
 		}
 
 		// Length must be specified for BLOB and TEXT column indexes.
 		if types.IsTypeBlob(col.FieldType.Tp) && ic.Length == types.UnspecifiedLength {
-			return nil, errors.Trace(errBlobKeyWithoutLength)
+			return nil, errBlobKeyWithoutLength
```

Comment:
@idlesummerbreeze 
Some errors forgot to traced in old code, some errors doesn't trace for performance.
But most of the time, add trace is better.